plugins/in_tail/tail_file.c

```diff
@@ -38,25 +38,201 @@ static inline void consume_bytes(char *buf, int bytes, int length)
+        else {
+            msgpack_pack_false(pck);
+        }
+        break;
+    case MSGPACK_OBJECT_POSITIVE_INTEGER:
+        msgpack_pack_uint64(pck, root->via.u64);
+        break;
+    case MSGPACK_OBJECT_NEGATIVE_INTEGER:
+        msgpack_pack_int64(pck, root->via.i64);
+        break;
+    case MSGPACK_OBJECT_FLOAT64:
+        msgpack_pack_double(pck, root->via.f64);
+        break;
+    case MSGPACK_OBJECT_FLOAT32:
+        msgpack_pack_float(pck, (float)root->via.f64);
+        break;
+    case MSGPACK_OBJECT_EXT:
+        {
+            int ret = msgpack_pack_ext(pck,
+                                       root->via.ext.size,
+                                       root->via.ext.type);
+            if (ret < 0) {
+                return ret;
+            }
+            msgpack_pack_ext_body(pck,
+                                  root->via.ext.ptr,
+                                  root->via.ext.size);
+        }
+        break;
+
+    case MSGPACK_OBJECT_STR:
+        {
+            int ret = msgpack_pack_str(pck, root->via.str.size);
+            if (ret < 0) {
+                return ret;
+            }
+            msgpack_pack_str_body(pck,
+                                  root->via.str.ptr,
+                                  root->via.str.size);
+        }
+        break;
+    case MSGPACK_OBJECT_BIN:
+        {
+            int ret = msgpack_pack_bin(pck, root->via.bin.size);
+            if (ret < 0) {
+                return ret;
+            }
+            msgpack_pack_bin_body(pck,
+                                  root->via.bin.ptr,
+                                  root->via.bin.size);
+        }
+        break;
+    case MSGPACK_OBJECT_ARRAY:
+        loop = root->via.array.size;
+        msgpack_pack_array(pck, loop);        
+        if (loop != 0){
+            msgpack_object* p = root->via.array.ptr;
+            for (i=0; i<loop; i++) {
+                unpack_and_pack(pck, p+i, first_map,
+                            key, key_len, val, val_len);
+            }
+        }
+        break;
+
+    case MSGPACK_OBJECT_MAP:
+        loop = root->via.map.size;
+        if (*first_map == FLB_TRUE) {
+            /* append path_key */
+            *first_map = FLB_FALSE;
+            msgpack_pack_map(pck, loop+1);
+            msgpack_pack_str(pck, key_len);
+            msgpack_pack_str_body(pck, key, key_len);
+            msgpack_pack_str(pck, val_len);
+            msgpack_pack_str_body(pck, val, val_len);
+        }
+        else {
+            msgpack_pack_map(pck, loop);
+        }
+        if (loop != 0) {
+            msgpack_object_kv *p = root->via.map.ptr;
+            for(i=0; i<loop; i++){
+                unpack_and_pack(pck, &(p+i)->key, first_map,
+                                key, key_len, val, val_len);
+                unpack_and_pack(pck, &(p+i)->val, first_map,
+                                key, key_len, val, val_len);
+            }
+        }
+        break;
+    }
+    return 0;
+}
```

Comment:
Thanks. I think it is specific for in_tail.
By the way, we try to implement unified schema to add custom fields.
So, it is useful to implement generic function.

How about cut off `unpack_and_pack` and expose it,  so called `flb_unpack_and_pack` ?